dist.Dockerfile

```diff
@@ -0,0 +1,46 @@
+FROM python:3-alpine as builder
```

Comment:
What is the difference between using this file versus our current Dockerfile? https://github.com/lyft/cartography/blob/master/Dockerfile

Reply:
I've tried to create a _securer_, _productionised_ Docker image here following the principles of [distroless images](https://github.com/GoogleContainerTools/distroless#why-should-i-use-distroless-images), and [some best practices published by hexops](https://github.com/hexops/dockerfile). I think this is a useful distinction to have between the current `./Dockerfile` which was initially implemented as a testing/development image in #606. 

The differences between this `./dist.Dockerfile` and `./Dockerfile` are:
 
1. We build a self-contained cartography binary using `pyinstaller`. Dependencies like `psutil` and `cryptography` need to be compiled which is why we install build tools like `gcc` and development headers for `openssl` etc..
2. We copy this self-contained cartography binary into a fresh [Alpine Linux](https://github.com/docker-library/docs/tree/master/alpine#what-is-alpine-linux) container which is commonly used as a minimal base OS. This results in a much smaller Docker image as unneeded compile-time dependencies aren't installed. Ideally, we'd love to use `scratch` but Python doesn't offer a way to statically compile into a single binary like Go or Rust do. Our self-contained binary from step 1. appears to depend on having `sh` and the same libc bindings that it was built with, so we're using Alpine Linux to both build and run cartography.
3. We configure the Docker container to run as a UID and GID that likely doesn't overlap with host users.

Reply:
It would be ideal to not use alpine. Its security is actually worse than using a slim python (ubuntu-based) image. There's a lot of reasons here, but the primary being:

* Alpine doesn't have a security vulnerability reporting process
* Its individual packages aren't stabilized and there's no way to report a vulnerability for them either (and they're community managed)

Past security, its use of musl is non-ideal because most python packages won't have wheels available, and we'll run into random incompatibilies due to musl itself.

The slim images, at this point, aren't much larger than alpine images, while having the good security teams/communities behind ubuntu and debian.

I'm fine with making some improvements to the existing Dockerfile, but ideally it would keep things simple and use a python-slim image and pip.

Reply:
Well, I guess a benefit of using pyinstaller is that it makes it a lot easier to use a multi-stage build like this. I guess I'm in favor of that change, but please use the python-slim image rather than alpine.

Reply:
To followup a bit, I also think it's best to keep the separate dockerfile, like this, in dist, if we're going to use pyinstaller, as it makes it more difficult to do quick development (like bind mounting the cartography code in, so changes don't require rebuilds).